File: 7-grafos/1200.py
class BinaryTree:
    def __init__(self,valueNode)->None:
        self.value=valueNode
        self.left=None
        self.right=None
        self.display=None
    def appendElement(self,node,newValue)->None:
        if not node.left and newValue<node.value+1:
            node.left=BinaryTree(newValue)
            return
        if not node.right and newValue>node.value:
            node.right=BinaryTree(newValue)
            return
        if newValue<node.value+1:node.appendElement(node.left,newValue)
        elif newValue>node.value:node.appendElement(node.right,newValue)
    def searchItem(self,node,sValue)->bool:
        if not node: return 0
        if node.value==sValue:return 1
        result=self.searchItem(node.left,sValue) if sValue<node.value+1 else self.searchItem(node.right,sValue)
        return result
    def preOrder(self,node)->None:
        if not node:return
        self.display.append(node.value)
        self.preOrder(node.left)
        self.preOrder(node.right)
    def inOrder(self,node)->None:
        if not node:return
        self.inOrder(node.left)
        self.display.append(node.value)
        self.inOrder(node.right)
    def postOrder(self,node)->None:
        if not node:return
        self.postOrder(node.left)
        self.postOrder(node.right)
        self.display.append(node.value)
    def displayBT(self,node,t):
        t=t[0]
        self.display=[]
        self.preOrder(node)if t=="PREFIXA" else self.inOrder(node)if t=="INFIXA" else self.postOrder(node)
        for i in range(len(self.display)-1):print(chr(self.display[i]),end=' ')
        print(chr(self.display[-1]))
BT=BinaryTree(0)
while 1:
    try:
        operation=input().split()
        if len(operation)<2:BT.displayBT(BT,operation)
        elif operation[0]=='I':
            if not BT.value:BT.value=ord(operation[1])
            else: BT.appendElement(BT,ord(operation[1]))
        elif BT.searchItem(BT,ord(operation[1])):print('{} existe'.format(operation[1]))
        else: print('{} nao existe'.format(operation[1]))
    except EOFError:break

File: 7-grafos/1200.cpp
#include<iostream>
#include<cstring>
#include<vector>
//
struct BST{
	int value;
	struct BST*right;
	struct BST*left;
	BST(int newValue){
		value=newValue;
		right=NULL;
		left=NULL;
	}
};
BST*insertBT(BST*ref,int value){
	if(!ref)return new BST(value);
	if(value<ref->value+1)ref->left=insertBT(ref->left,value);
	else ref->right=insertBT(ref->right,value);
	return ref;
}
int searchBT(BST*ref,int value){
	int result=0;
	if(!ref)return 0;
	if(ref->value==value)return 1;
	if(value>ref->value)result=searchBT(ref->right,value);
	else result=searchBT(ref->left,value);
	return result;
}
//
std::vector<int>display;
void Prefix(BST*ref){
	if(!ref)return;
	display.push_back(ref->value);
	Prefix(ref->left);
	Prefix(ref->right);
}
void Infix(BST*ref){
	if(!ref)return;
	Infix(ref->left);
	display.push_back(ref->value);
	Infix(ref->right);
}
void Posfix(BST*ref){
	if(!ref)return;
	Posfix(ref->left);
	Posfix(ref->right);
	display.push_back(ref->value);
}
void displayBST(BST*bTree,std::string type){
	if(type=="INFIXA")Infix(bTree);
	else if(type=="PREFIXA")Prefix(bTree);
	else Posfix(bTree);
	for(int i=0;i<display.size()-1;++i)std::cout<<(char)display[i]<<" ";
	std::cout<<(char)display.back()<<"\n";
	display.clear();
}
//
void commandInsert(BST*&bTree,int nValue){
	if(!bTree)bTree=new BST(nValue);
	else bTree=insertBT(bTree,nValue);
}
//
int main(){
	std::string inp;
	BST*binaryTree=NULL;
	while(std::getline(std::cin,inp)&&!std::cin.eof()){
		std::string extractInps[2]={" "," "};
		char*split=std::strtok(&inp[0]," ");
		for(int i=0;split!=NULL;++i){
			extractInps[i]=split;
			split=std::strtok(NULL," ");
		}
		if(extractInps[1]==" ")displayBST(binaryTree,extractInps[0]);
		else if(extractInps[0]=="I")commandInsert(binaryTree,(int)extractInps[1][0]);
		else if(!searchBT(binaryTree,(int)extractInps[1][0]))std::cout<<extractInps[1]<<" nao existe\n";
		else std::cout<<extractInps[1]<<" existe\n";
	}
	return 0;
}
